### PR TITLE
resource/aws_lightsail_instance: Add validation for instance name

### PR DESCRIPTION
**Pull request by nywilken**
related to #8647

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

Fixes #0000

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
None
```

Output from acceptance testing:

```
--- PASS: TestAccAWSLightsailInstance_Name (99.68s)
```

### CODE REVIEW

File: aws/resource_aws_lightsail_instance.go [hunk 1]
```diff
@@ -24,9 +26,10 @@ func resourceAwsLightsailInstance() *schema.Resource {
 
 		Schema: map[string]*schema.Schema{
 			"name": {
-				Type:     schema.TypeString,
-				Required: true,
-				ForceNew: true,
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringMatch(regexp.MustCompile(`^[[:alnum:]]{2}[[:alnum:]_\-]*$`), "must be at least 2 characters long and contain only alphanumerics, hyphens, and dashes"),
```
Comment (paddycarver) [inline aws/resource_aws_lightsail_instance.go]: Is `a-123` a valid name? I believe this regex wouldn't match that.

Comment (bflad) [reply aws/resource_aws_lightsail_instance.go]: My personal preference for more complex validation functions nowadays is utilizing [`validation.All()`](https://godoc.org/github.com/hashicorp/terraform/helper/validation#All) so multiple restrictions can be broken down into easier functions. Certainly not a requirement though, just thought I would mention!

```suggestion
				ValidateFunc: validation.All(
					validation.StringLenBetween(3, 255),
					validation.StringMatch(regexp.MustCompile(`^[a-zA-Z0-9_-]*$`), "must contain only alphanumeric characters, underscores, and hyphens"),
				),
```

My hope is that we'll eventually find emergent patterns with `StringMatch` and create wrappers/constants to simplify this even further.

If you are curious, I was able to coerce the max length error with [`acctest.RandString()`](https://godoc.org/github.com/hashicorp/terraform/helper/acctest#RandString) 😄 

```go
lightsailName := acctest.RandString(1025)
```

```
--- FAIL: TestAccAWSLightsailInstance_basic (7.34s)
    testing.go:568: Step 0 error: errors during apply:

        Error: InvalidInputException: 81gjehvhlkp8zb809x3bsb7zuoxnqtcxwu4hqdt8zlbhqyo8qockfumus0jgr6cyae4unytj3ms7okj2yo6z2ksnyw4ws9bv2n0tirk9prhfmuntcjr9z1eazufbtlttrxwd7owds2fwt3pxnxaanwwhhrfmf339apqh8dtci4zg99ocmzarxjbhhhymtndtof0u0101s3rfevonb0br1196srrpb82cir8fg8unxjgy8ntzflwq7xwkpbunr33vh76jj4voe3he36rz20v9iktvqh63ha2wj87pmndpp0uyvefb9j1s89q9zfv3aifnywc0el66yx4czfzzyc9xaq6rdebhczpe7bat6apsdmazhqt139ql7rhtlauld3q4df3d9cncf3fad7o8nrjlzy3ow6mxmuwgi8bbg1fwr27trf8imx9yroh1f82psmywon3bai1irb93krgb3lu73zuj0z1mnb0sofmq7yayid41rypqpkxwrbs3od64ia7j0s34z941f604wd606d6huvxnvgcp6nxpjqvw0mgyjw9ndrmnczkmfa06d0sb64yhm6hrk3wbako7prgarub1ycnlsfmrx1kas6xd69xu26p0va3o1eot7f0nkec2vjftycs713p8hztnnrfjwf29xlx7gomx6r80duoxbjkb1rgek8om8ex3xcg3wy2p6s0dwpuhg1oakyqrmjj0j7n6gqvs3wpsshbycq1rxr3dcjylhfbqgorzqt6ugz1b8map763v0cd0bqjigadwlemsf1gd9gvlu0ax7wbamlnshqxgowebhllopcjbhdcml76v6mhqld01gw231p8349nrr7fcd8aidwd0kn4xwl34f0jh0s4eerh7z1vr1bdwe0xpzctdjaugrhgy81xr3q7r6j2sd3ttvsh3jyj7xuty8nqd2vhxgcj1qke0s2cihdoxnkau4quwsx0a72w8odesbolcmhbtm3qmjday97t0y99s27f0sm2rt4i3mpjn7gmkp is not a valid resource name. Length exceeds maximum of 255
```

Comment (nywilken) [reply aws/resource_aws_lightsail_instance.go]: @paddycarver good catch :bowing_man:  

@bflad no one said anything about a limit :stuck_out_tongue_winking_eye: Good catch as well. I'll address both cases and take a look at `validation.All`. Although, I do enjoy looking at a complex validation every now and then.

Comment (nywilken) [reply aws/resource_aws_lightsail_instance.go]: I actually like the `validation.All` approach :heavy_check_mark:

**Comment by nywilken** [comment]
@paddycarver @bflad if you could take another look that would be great. I opted to double up on the error message for each validation so that a user can quickly see what is considered a valid name as opposed to see the error in parts.

File: aws/resource_aws_lightsail_instance.go (hunk 1 above)
Comment (paddycarver) [reply aws/resource_aws_lightsail_instance.go]: My concern is resolved, but I defer to @bflad.

**Comment by bflad** [comment]
@nywilken sorry for moving this one back -- I would like to show you some things here. 😄

**Comment by nywilken** [comment]
> @nywilken sorry for moving this one back -- I would like to show you some things here. smile

:+1: looking forward to it, thanks.

**Comment by bflad** [comment]
This has been released in [version 2.19.0 of the Terraform AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#2190-july-11-2019). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

For further feature requests or bug reports with this functionality, please create a [new GitHub issue](https://github.com/terraform-providers/terraform-provider-aws/issues/new/choose) following the template for triage. Thanks!

**Pull request merged**